File: services/backend/app/services/marketplace.py
"""
Marketplace Service — Browse, search, install agents, reviews, categories, featured.

Integrates with MarketplaceListing, MarketplaceReview, MarketplaceInstall models
and the AgentService for duplicating agents on install.
"""

from __future__ import annotations

import uuid
from datetime import datetime, timezone
from typing import Any

from sqlalchemy import and_, case, func, or_, select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import joinedload
from structlog import get_logger

from app.core.config import settings
from app.core.events import get_session_factory
from app.core.exceptions import (
    ConflictError,
    NotFoundError,
    ValidationError,
)
from app.models.agent import Agent
from app.models.marketplace import (
    MarketplaceListing,
    MarketplaceReview,
    MarketplaceInstall,
)
from app.models.user import User
from app.services.agent import AgentService

logger = get_logger(__name__)

# Valid sort options for marketplace listing
VALID_SORTS = ("popular", "newest", "rating", "price_low", "price_high")


class MarketplaceService:
    """Business logic for the Anansi marketplace (The Bazaar)."""

    def __init__(self) -> None:
        self._agent_service: AgentService | None = None

    @property
    def agent_service(self) -> AgentService:
        if self._agent_service is None:
            self._agent_service = AgentService()
        return self._agent_service

    # ── Browse ─────────────────────────────────────────────────────────────────

    async def list_listings(
        self,
        category: str | None = None,
        tags: list[str] | None = None,
        sort: str = "popular",
        page: int = 1,
        limit: int = 20,
        price_min: int | None = None,
        price_max: int | None = None,
    ) -> dict[str, Any]:
        """Browse marketplace listings with filters and pagination.

        Args:
            category: Filter by category slug.
            tags: Filter by tags (AND logic).
            sort: Sort order — 'popular', 'newest', 'rating', 'price_low', 'price_high'.
            page: Page number (1-indexed).
            limit: Items per page (max 50).
            price_min: Minimum price in cents (0 = free).
            price_max: Maximum price in cents.

        Returns:
            Dict with 'items', 'total', 'page', 'limit', 'pages'.
        """
        if sort not in VALID_SORTS:
            raise ValidationError(
                message=f"Invalid sort option. Must be one of: {', '.join(VALID_SORTS)}"
            )

        factory = get_session_factory()
        async with factory() as session:
            # Base query: only published listings
            query = (
                select(MarketplaceListing)
                .options(
                    joinedload(MarketplaceListing.agent),
                    joinedload(MarketplaceListing.reviews),
                )
                .where(MarketplaceListing.status == "published")
            )

            # ── Filters ──
            if category:
                query = query.where(MarketplaceListing.category == category)

            if tags:
                # PostgreSQL ARRAY overlap operator: && (any overlap)
                # Using ANY for simple tag filtering
                for tag in tags:
                    query = query.where(
                        func.array_position(MarketplaceListing.tags, tag) is not None
                    )

            if price_min is not None:
                query = query.where(MarketplaceListing.price_cents >= price_min)

            if price_max is not None:
                query = query.where(MarketplaceListing.price_cents <= price_max)

            # ── Sort ──
            if sort == "popular":
                query = query.order_by(MarketplaceListing.install_count.desc())
            elif sort == "newest":
                query = query.order_by(MarketplaceListing.created_at.desc())
            elif sort == "rating":
                query = query.order_by(MarketplaceListing.rating_avg.desc())
            elif sort == "price_low":
                query = query.order_by(MarketplaceListing.price_cents.asc())
            elif sort == "price_high":
                query = query.order_by(MarketplaceListing.price_cents.desc())

            # ── Count ──
            count_query = select(func.count()).select_from(query.subquery())
            total_result = await session.execute(count_query)
            total = total_result.scalar() or 0

            # ── Paginate ──
            limit = min(limit, 50)
            pages = max(1, (total + limit - 1) // limit)
            offset = (page - 1) * limit

            query = query.offset(offset).limit(limit)

            result = await session.execute(query)
            listings = result.unique().scalars().all()

            return {
                "items": [self._serialize_listing(l) for l in listings],
                "total": total,
                "page": page,
                "limit": limit,
                "pages": pages,
            }

    # ── Detail ─────────────────────────────────────────────────────────────────

    async def get_listing(self, listing_id: str) -> dict[str, Any]:
        """Get full marketplace listing detail with screenshots, rating, creator info.

        Args:
            listing_id: UUID of the marketplace listing.

        Returns:
            Full listing detail dict.

        Raises:
            NotFoundError: If listing doesn't exist or isn't published.
        """
        factory = get_session_factory()
        async with factory() as session:
            listing = await self._fetch_listing(session, listing_id)

            # Get creator info
            creator_query = select(User).where(User.id == listing.user_id)
            creator_result = await session.execute(creator_query)
            creator = creator_result.scalar_one_or_none()

            # Get version history from the agent
            agent = listing.agent
            version_history = []
            if agent:
                versions = await self.agent_service.get_versions(str(agent.id))
                version_history = [
                    {
                        "version": v["version"],
                        "created_at": v["created_at"],
                        "change_notes": v.get("change_notes"),
                    }
                    for v in versions[-10:]  # Last 10 versions
                ]

            return {
                **self._serialize_listing(listing),
                "version_history": version_history,
                "creator": {
                    "id": str(creator.id) if creator else None,
                    "display_name": creator.display_name if creator else "Unknown",
                    "avatar_url": creator.avatar_url if creator else None,
                },
                "reviews": [
                    self._serialize_review(r) for r in listing.reviews
                ],
            }

    # ── Install ────────────────────────────────────────────────────────────────

    async def install_agent(self, listing_id: str, user_id: str) -> dict[str, Any]:
        """Install a marketplace agent — creates a copy in the user's library.

        Steps:
        1. Verify the listing exists and is published.
        2. Check user hasn't already installed this listing.
        3. Record the install.
        4. Increment the listing's install_count.
        5. Duplicate the agent under the user's ownership.

        Args:
            listing_id: UUID of the marketplace listing.
            user_id: UUID of the installing user.

        Returns:
            Dict with the new agent info and install record.

        Raises:
            NotFoundError: Listing not found.
            ConflictError: Already installed.
        """
        factory = get_session_factory()
        async with factory() as session:
            listing = await self._fetch_listing(session, listing_id)

            # Check for existing install
            existing_check = await session.execute(
                select(MarketplaceInstall).where(
                    and_(
                        MarketplaceInstall.listing_id == listing.id,
                        MarketplaceInstall.user_id == uuid.UUID(user_id),
                    )
                )
            )
            if existing_check.scalar_one_or_none():
                raise ConflictError(
                    message="You have already installed this agent",
                    details={"listing_id": listing_id},
                )

            # Duplicate the agent for the user
            agent = listing.agent
            if not agent:
                raise NotFoundError(
                    message="The source agent for this listing no longer exists",
                    resource_type="agent",
                )

            # Create a copy of the agent definition
            from copy import deepcopy

            new_agent = Agent(
                user_id=uuid.UUID(user_id),
                name=agent.name,
                description=agent.description,
                definition=deepcopy(agent.definition),
                version=1,
                status="active",
                is_published=False,
                category=agent.category,
                tags=agent.tags[:] if agent.tags else [],
            )
            session.add(new_agent)
            await session.flush()  # Get the ID

            # Record the install
            install = MarketplaceInstall(
                listing_id=listing.id,
                user_id=uuid.UUID(user_id),
                agent_id=new_agent.id,
            )
            session.add(install)

            # Increment install count
            listing.install_count = (listing.install_count or 0) + 1

            await session.commit()

            logger.info(
                "Agent installed from marketplace",
                listing_id=listing_id,
                user_id=user_id,
                new_agent_id=str(new_agent.id),
            )

            return {
                "install_id": str(install.id),
                "agent": {
                    "id": str(new_agent.id),
                    "name": new_agent.name,
                    "description": new_agent.description,
                    "version": new_agent.version,
                },
                "listing_title": listing.title,
                "installed_at": install.created_at.isoformat() if install.created_at else None,
            }

    # ── Search ─────────────────────────────────────────────────────────────────

    async def search_listings(
        self,
        query: str,
        category: str | None = None,
        tags: list[str] | None = None,
        price_free: bool | None = None,
        page: int = 1,
        limit: int = 20,
    ) -> dict[str, Any]:
        """Full-text search across marketplace listings.

        Searches title, description, and tags using PostgreSQL full-text search.

        Args:
            query: Search query string.
            category: Optional category filter.
            tags: Optional tags filter.
            price_free: If True, show only free agents. If False, show only paid.
            page: Page number (1-indexed).
            limit: Items per page (max 50).

        Returns:
            Dict with 'items', 'total', 'page', 'limit', 'pages'.
        """
        if not query or not query.strip():
            raise ValidationError(message="Search query is required")

        factory = get_session_factory()
        async with factory() as session:
            search_term = f"%{query.strip()}%"

            base_query = (
                select(MarketplaceListing)
                .options(joinedload(MarketplaceListing.agent))
                .where(
                    and_(
                        MarketplaceListing.status == "published",
                        or_(
                            MarketplaceListing.title.ilike(search_term),
                            MarketplaceListing.description.ilike(search_term),
                            func.array_to_string(MarketplaceListing.tags, " ").ilike(
                                search_term
                            ),
                        ),
                    )
                )
            )

            if category:
                base_query = base_query.where(
                    MarketplaceListing.category == category
                )

            if tags:
                for tag in tags:
                    base_query = base_query.where(
                        func.array_position(MarketplaceListing.tags, tag) is not None
                    )

            if price_free is True:
                base_query = base_query.where(MarketplaceListing.price_cents == 0)
            elif price_free is False:
                base_query = base_query.where(MarketplaceListing.price_cents > 0)

            # Count
            count_query = select(func.count()).select_from(base_query.subquery())
            total_result = await session.execute(count_query)
            total = total_result.scalar() or 0

            # Paginate
            limit = min(limit, 50)
            pages = max(1, (total + limit - 1) // limit)
            offset = (page - 1) * limit

            # Sort by relevance (title matches first, then description matches)
            title_match = case(
                (MarketplaceListing.title.ilike(search_term), 0),
                else_=1,
            )
            base_query = base_query.order_by(title_match, MarketplaceListing.install_count.desc())
            base_query = base_query.offset(offset).limit(limit)

            result = await session.execute(base_query)
            listings = result.unique().scalars().all()

            return {
                "items": [self._serialize_listing(l) for l in listings],
                "total": total,
                "page": page,
                "limit": limit,
                "pages": pages,
            }

    # ── Featured ───────────────────────────────────────────────────────────────

    async def get_featured(self, limit: int = 10) -> list[dict[str, Any]]:
        """Get curated/featured marketplace listings.

        Returns listings with is_featured=True, sorted by install count.

        Args:
            limit: Maximum number of featured listings to return.

        Returns:
            List of serialized featured listings.
        """
        factory = get_session_factory()
        async with factory() as session:
            query = (
                select(MarketplaceListing)
                .options(joinedload(MarketplaceListing.agent))
                .where(
                    and_(
                        MarketplaceListing.status == "published",
                        MarketplaceListing.is_featured == True,
                    )
                )
                .order_by(MarketplaceListing.install_count.desc())
                .limit(limit)
            )
            result = await session.execute(query)
            listings = result.unique().scalars().all()

            return [self._serialize_listing(l) for l in listings]

    # ── Categories ─────────────────────────────────────────────────────────────

    async def get_categories(self) -> list[dict[str, Any]]:
        """List all categories with listing counts.

        Returns:
            List of dicts with 'category', 'count', and 'display_name'.
        """
        CATEGORY_DISPLAY_NAMES = {
            "productivity": "Productivity",
            "communications": "Communications",
            "finance": "Finance",
            "social_media": "Social Media",
            "devops": "DevOps",
            "ecommerce": "E-commerce",
            "health": "Health",
            "education": "Education",
            "personal": "Personal",
            "custom": "Custom",
            "uncategorized": "Uncategorized",
        }

        factory = get_session_factory()
        async with factory() as session:
            query = (
                select(
                    MarketplaceListing.category,
                    func.count(MarketplaceListing.id).label("count"),
                )
                .where(
                    and_(
                        MarketplaceListing.status == "published",
                        MarketplaceListing.category.isnot(None),
                    )
                )
                .group_by(MarketplaceListing.category)
                .order_by(func.count(MarketplaceListing.id).desc())
            )
            result = await session.execute(query)
            rows = result.all()

            categories = []
            for row in rows:
                slug = row.category
                categories.append({
                    "slug": slug,
                    "display_name": CATEGORY_DISPLAY_NAMES.get(
                        slug, slug.replace("_", " ").title()
                    ),
                    "count": row.count,
                })

            return categories

    # ── Reviews ────────────────────────────────────────────────────────────────

    async def submit_review(
        self,
        listing_id: str,
        user_id: str,
        rating: int,
        review: str | None = None,
    ) -> dict[str, Any]:
        """Submit or update a review for a marketplace listing.

        One review per user per listing — if a review exists, it is updated.

        Args:
            listing_id: UUID of the marketplace listing.
            user_id: UUID of the reviewer.
            rating: Rating 1-5.
            review: Optional review text.

        Returns:
            Serialized review dict.

        Raises:
            ValidationError: If rating is out of range.
            NotFoundError: If listing doesn't exist.
        """
        if rating < 1 or rating > 5:
            raise ValidationError(message="Rating must be between 1 and 5")

        factory = get_session_factory()
        async with factory() as session:
            listing = await self._fetch_listing(session, listing_id)

            # Check for existing review
            existing_query = select(MarketplaceReview).where(
                and_(
                    MarketplaceReview.listing_id == listing.id,
                    MarketplaceReview.user_id == uuid.UUID(user_id),
                )
            )
            existing_result = await session.execute(existing_query)
            existing_review = existing_result.scalar_one_or_none()

            old_rating = None
            if existing_review:
                old_rating = existing_review.rating
                existing_review.rating = rating
                existing_review.review = review
            else:
                new_review = MarketplaceReview(
                    listing_id=listing.id,
                    user_id=uuid.UUID(user_id),
                    rating=rating,
                    review=review,
                )
                session.add(new_review)

            # Recalculate average rating
            await self._recalculate_rating(session, listing, old_rating, rating)
            await session.commit()

            # Refresh to get updated listing
            await session.refresh(listing)

            # Return the review (existing or new)
            review_query = select(MarketplaceReview).where(
                and_(
                    MarketplaceReview.listing_id == listing.id,
                    MarketplaceReview.user_id == uuid.UUID(user_id),
                )
            )
            review_result = await session.execute(review_query)
            final_review = review_result.scalar_one()

            return {
                "id": str(final_review.id),
                "listing_id": listing_id,
                "user_id": user_id,
                "rating": final_review.rating,
                "review": final_review.review,
                "created_at": final_review.created_at.isoformat() if final_review.created_at else None,
                "listing_rating_avg": listing.rating_avg,
                "listing_rating_count": listing.rating_count,
            }

    # ── Internal Helpers ───────────────────────────────────────────────────────

    @staticmethod
    async def _recalculate_rating(
        session: AsyncSession,
        listing: MarketplaceListing,
        old_rating: int | None,
        new_rating: int,
    ) -> None:
        """Recalculate the average rating for a listing after a review change."""
        # Get all ratings for this listing
        ratings_query = select(MarketplaceReview.rating).where(
            MarketplaceReview.listing_id == listing.id
        )
        ratings_result = await session.execute(ratings_query)
        ratings = [r[0] for r in ratings_result.all()]

        # Handle the case where the review being updated is already counted
        # by re-reading from DB after flush
        await session.flush()
        ratings_query2 = select(MarketplaceReview.rating).where(
            MarketplaceReview.listing_id == listing.id
        )
        ratings_result2 = await session.execute(ratings_query2)
        all_ratings = [r[0] for r in ratings_result2.all()]

        if all_ratings:
            listing.rating_avg = round(sum(all_ratings) / len(all_ratings), 2)
            listing.rating_count = len(all_ratings)
        else:
            listing.rating_avg = 0.0
            listing.rating_count = 0

    @staticmethod
    async def _fetch_listing(
        session: AsyncSession,
        listing_id: str,
    ) -> MarketplaceListing:
        """Fetch a marketplace listing by ID, raising NotFoundError if missing."""
        try:
            uid = uuid.UUID(listing_id)
        except ValueError as exc:
            raise ValidationError(message=f"Invalid listing ID: {listing_id}") from exc

        result = await session.execute(
            select(MarketplaceListing)
            .options(
                joinedload(MarketplaceListing.agent),
                joinedload(MarketplaceListing.reviews),
                joinedload(MarketplaceListing.installs),
            )
            .where(MarketplaceListing.id == uid)
        )
        listing = result.unique().scalar_one_or_none()

        if listing is None:
            raise NotFoundError(
                message=f"Marketplace listing not found: {listing_id}",
                resource_type="marketplace_listing",
                resource_id=listing_id,
            )

        return listing

    @staticmethod
    def _serialize_listing(listing: MarketplaceListing) -> dict[str, Any]:
        """Convert a MarketplaceListing ORM instance to a safe API response dict."""
        agent = listing.agent

        return {
            "id": str(listing.id),
            "agent_id": str(listing.agent_id) if listing.agent_id else None,
            "user_id": str(listing.user_id),
            "title": listing.title,
            "description": listing.description,
            "price_cents": listing.price_cents,
            "price_display": (
                f"${listing.price_cents / 100:.2f}" if listing.price_cents > 0 else "Free"
            ),
            "category": listing.category,
            "tags": listing.tags or [],
            "screenshots": listing.screenshots or [],
            "rating_avg": listing.rating_avg,
            "rating_count": listing.rating_count,
            "install_count": listing.install_count,
            "status": listing.status,
            "is_featured": listing.is_featured,
            "agent_name": agent.name if agent else None,
            "agent_version": agent.version if agent else None,
            "memory_nodes_per_run": listing.memory_nodes_per_run,
            "memory_links_per_run": listing.memory_links_per_run,
            "rejection_reason": listing.rejection_reason,
            "created_at": listing.created_at.isoformat() if listing.created_at else None,
            "updated_at": listing.updated_at.isoformat() if listing.updated_at else None,
        }

    @staticmethod
    def _serialize_review(review: MarketplaceReview) -> dict[str, Any]:
        """Serialize a single review."""
        return {
            "id": str(review.id),
            "listing_id": str(review.listing_id),
            "user_id": str(review.user_id),
            "rating": review.rating,
            "review": review.review,
            "created_at": review.created_at.isoformat() if review.created_at else None,
        }


__all__ = ["MarketplaceService"]

File: anansi-frontend-fix/src/stores/marketplace.ts
/**
 * Marketplace Zustand Store — Browse, search, install, reviews, categories.
 *
 * Manages marketplace state: listings, search, filters, categories, featured,
 * installation flow, and reviews.
 */

import { create } from "zustand";
import { api } from "../lib/api";

// ─── Types ──────────────────────────────────────────────────────────────────

export interface MarketplaceListing {
  id: string;
  agent_id: string | null;
  user_id: string;
  title: string;
  description: string | null;
  price_cents: number;
  price_display: string;
  category: string | null;
  tags: string[];
  screenshots: string[];
  rating_avg: number;
  rating_count: number;
  install_count: number;
  status: string;
  is_featured: boolean;
  agent_name: string | null;
  agent_version: number | null;
  memory_nodes_per_run: number | null;
  memory_links_per_run: number | null;
  rejection_reason: string | null;
  created_at: string | null;
  updated_at: string | null;
}

export interface MarketplaceListingDetail extends MarketplaceListing {
  version_history: {
    version: number;
    created_at: string;
    change_notes: string | null;
  }[];
  creator: {
    id: string | null;
    display_name: string;
    avatar_url: string | null;
  };
  reviews: MarketplaceReview[];
}

export interface MarketplaceReview {
  id: string;
  listing_id: string;
  user_id: string;
  rating: number;
  review: string | null;
  created_at: string | null;
  reviewer_name?: string;
  listing_title?: string;
}

export interface MarketplaceCategory {
  slug: string;
  display_name: string;
  count: number;
}

export interface PaginatedResponse {
  items: MarketplaceListing[];
  total: number;
  page: number;
  limit: number;
  pages: number;
}

// ─── Filter / Sort Types ──────────────────────────────────────────────────

export type SortOption = "popular" | "newest" | "rating" | "price_low" | "price_high";

export interface MarketplaceFilters {
  category: string | null;
  tags: string[];
  sort: SortOption;
  price_min: number | null;
  price_max: number | null;
  price_free: boolean | null;
  search: string;
}

// ─── Creator Dashboard Types ──────────────────────────────────────────────

export interface CreatorListing {
  id: string;
  agent_id: string | null;
  title: string;
  description: string | null;
  price_cents: number;
  price_display: string;
  category: string | null;
  status: string;
  is_featured: boolean;
  rating_avg: number;
  rating_count: number;
  install_count: number;
  is_published: boolean;
  estimated_earnings_cents: number;
  agent_name: string | null;
  created_at: string | null;
  updated_at: string | null;
}

export interface SalesAnalytics {
  period: string;
  total_installs: number;
  total_installs_all_time: number;
  total_reviews: number;
  average_rating: number;
  total_revenue_cents: number;
  total_platform_fees_cents: number;
  top_listing: {
    title: string | null;
    installs: number;
    price_cents: number;
  } | null;
  installs_over_time: {
    date: string;
    installs: number;
  }[];
}

export interface EarningsData {
  total_revenue_cents: number;
  total_earnings_cents: number;
  total_earnings_display: string;
  platform_fees_cents: number;
  platform_fees_display: string;
  pending_payout_cents: number;
  pending_payout_display: string;
  total_installs: number;
  paid_installs: number;
  free_installs: number;
  listings_count: number;
  payout_history: {
    id: string;
    amount_cents: number;
    amount_display: string;
    status: string;
    paid_at: string;
    period: string;
  }[];
  revenue_share: {
    creator_percentage: number;
    platform_percentage: number;
  };
}

// ─── Store Interface ─────────────────────────────────────────────────────

interface MarketplaceState {
  // Listing data
  listings: MarketplaceListing[];
  totalListings: number;
  currentPage: number;
  totalPages: number;

  // Detail
  selectedListing: MarketplaceListingDetail | null;

  // Categories
  categories: MarketplaceCategory[];
  featuredListings: MarketplaceListing[];

  // Filters
  filters: MarketplaceFilters;

  // Loading states
  isLoading: boolean;
  isInstalling: boolean;
  isSubmittingReview: boolean;
  error: string | null;

  // Creator dashboard
  myListings: CreatorListing[];
  analytics: SalesAnalytics | null;
  earnings: EarningsData | null;
  creatorReviews: MarketplaceReview[];

  // Actions
  fetchListings: (page?: number) => Promise<void>;
  fetchListingDetail: (id: string) => Promise<void>;
  installAgent: (listingId: string) => Promise<{ agent: { id: string; name: string } } | null>;
  submitReview: (listingId: string, rating: number, review?: string) => Promise<void>;
  fetchCategories: () => Promise<void>;
  fetchFeatured: () => Promise<void>;
  searchListings: (query: string, page?: number) => Promise<void>;
  setFilters: (filters: Partial<MarketplaceFilters>) => void;
  clearFilters: () => void;

  // Creator actions
  fetchMyListings: () => Promise<void>;
  fetchAnalytics: (period?: string) => Promise<void>;
  fetchEarnings: () => Promise<void>;
  fetchCreatorReviews: () => Promise<void>;
}

// ─── Initial Filters ────────────────────────────────────────────────────

const DEFAULT_FILTERS: MarketplaceFilters = {
  category: null,
  tags: [],
  sort: "popular",
  price_min: null,
  price_max: null,
  price_free: null,
  search: "",
};

// ─── Store ─────────────────────────────────────────────────────────────

export const useMarketplaceStore = create<MarketplaceState>((set, get) => ({
  // ── State ──
  listings: [],
  totalListings: 0,
  currentPage: 1,
  totalPages: 0,
  selectedListing: null,
  categories: [],
  featuredListings: [],
  filters: { ...DEFAULT_FILTERS },
  isLoading: false,
  isInstalling: false,
  isSubmittingReview: false,
  error: null,

  myListings: [],
  analytics: null,
  earnings: null,
  creatorReviews: [],

  // ── Actions ──

  fetchListings: async (page?: number) => {
    const { filters } = get();
    set({ isLoading: true, error: null });

    try {
      const params: Record<string, string | number | boolean> = {
        page: page ?? get().currentPage,
        limit: 20,
        sort: filters.sort,
      };
      if (filters.category) params.category = filters.category;
      if (filters.price_min !== null) params.price_min = filters.price_min;
      if (filters.price_max !== null) params.price_max = filters.price_max;
      if (filters.tags.length > 0) params.tags = filters.tags.join(",");

      const data = await api.get<PaginatedResponse>("/api/v1/marketplace", { params });

      set({
        listings: data.items,
        totalListings: data.total,
        currentPage: data.page,
        totalPages: data.pages,
        isLoading: false,
      });
    } catch (err: unknown) {
      const message = err instanceof Error ? err.message : "Failed to load marketplace";
      set({ isLoading: false, error: message });
    }
  },

  fetchListingDetail: async (id: string) => {
    set({ isLoading: true, error: null, selectedListing: null });

    try {
      const data = await api.get<MarketplaceListingDetail>(`/api/v1/marketplace/${id}`);
      set({ selectedListing: data, isLoading: false });
    } catch (err: unknown) {
      const message = err instanceof Error ? err.message : "Failed to load listing";
      set({ isLoading: false, error: message });
    }
  },

  installAgent: async (listingId: string) => {
    set({ isInstalling: true, error: null });

    try {
      const data = await api.post<{ agent: { id: string; name: string } }>(
        `/api/v1/marketplace/${listingId}/install`
      );

      // Update local listing count
      const { listings, selectedListing } = get();
      set({
        listings: listings.map(l =>
          l.id === listingId ? { ...l, install_count: l.install_count + 1 } : l
        ),
        selectedListing: selectedListing?.id === listingId
          ? { ...selectedListing, install_count: selectedListing.install_count + 1 }
          : selectedListing,
        isInstalling: false,
      });

      return data;
    } catch (err: unknown) {
      const message = err instanceof Error ? err.message : "Failed to install agent";
      set({ isInstalling: false, error: message });
      return null;
    }
  },

  submitReview: async (listingId: string, rating: number, review?: string) => {
    set({ isSubmittingReview: true, error: null });

    try {
      const data = await api.post<MarketplaceReview>(
        `/api/v1/marketplace/${listingId}/review`,
        { rating, review }
      );

      // Reload detail to get updated reviews
      get().fetchListingDetail(listingId);
      set({ isSubmittingReview: false });
    } catch (err: unknown) {
      const message = err instanceof Error ? err.message : "Failed to submit review";
      set({ isSubmittingReview: false, error: message });
    }
  },

  fetchCategories: async () => {
    try {
      const data = await api.get<MarketplaceCategory[]>("/api/v1/marketplace/categories");
      set({ categories: data });
    } catch {
      // Non-critical, silently fail
    }
  },

  fetchFeatured: async () => {
    try {
      const data = await api.get<MarketplaceListing[]>("/api/v1/marketplace/featured");
      set({ featuredListings: data });
    } catch {
      // Non-critical
    }
  },

  searchListings: async (query: string, page?: number) => {
    set({ isLoading: true, error: null });

    try {
      const params: Record<string, string | number | boolean> = {
        q: query,
        page: page ?? 1,
        limit: 20,
      };

      const { filters } = get();
      if (filters.category) params.category = filters.category;
      if (filters.price_free !== null) params.price_free = filters.price_free;

      const data = await api.get<PaginatedResponse>("/api/v1/marketplace/search", { params });

      set({
        listings: data.items,
        totalListings: data.total,
        currentPage: data.page,
        totalPages: data.pages,
        isLoading: false,
      });
    } catch (err: unknown) {
      const message = err instanceof Error ? err.message : "Search failed";
      set({ isLoading: false, error: message });
    }
  },

  setFilters: (filters: Partial<MarketplaceFilters>) => {
    set(state => ({
      filters: { ...state.filters, ...filters },
      currentPage: 1, // Reset to first page on filter change
    }));
    // Auto-fetch with new filters
    get().fetchListings(1);
  },

  clearFilters: () => {
    set({ filters: { ...DEFAULT_FILTERS }, currentPage: 1 });
    get().fetchListings(1);
  },

  // ── Creator Actions ──

  fetchMyListings: async () => {
    try {
      const data = await api.get<CreatorListing[]>("/api/v1/marketplace/my-listings");
      set({ myListings: data });
    } catch (err: unknown) {
      const message = err instanceof Error ? err.message : "Failed to load listings";
      set({ error: message });
    }
  },

  fetchAnalytics: async (period?: string) => {
    try {
      const params: Record<string, string> = {};
      if (period) params.period = period;

      const data = await api.get<SalesAnalytics>("/api/v1/marketplace/analytics", { params });
      set({ analytics: data });
    } catch (err: unknown) {
      const message = err instanceof Error ? err.message : "Failed to load analytics";
      set({ error: message });
    }
  },

  fetchEarnings: async () => {
    try {
      const data = await api.get<EarningsData>("/api/v1/marketplace/earnings");
      set({ earnings: data });
    } catch (err: unknown) {
      const message = err instanceof Error ? err.message : "Failed to load earnings";
      set({ error: message });
    }
  },

  fetchCreatorReviews: async () => {
    try {
      const data = await api.get<MarketplaceReview[]>("/api/v1/marketplace/reviews");
      set({ creatorReviews: data });
    } catch (err: unknown) {
      const message = err instanceof Error ? err.message : "Failed to load reviews";
      set({ error: message });
    }
  },
}));
